front/components/data_source_view/DataSourceBuilderSelector.tsx

```diff
@@ -0,0 +1,242 @@
+import type { BreadcrumbItem } from "@dust-tt/sparkle";
+import { Breadcrumbs, SearchInput, Spinner } from "@dust-tt/sparkle";
+import type { Dispatch, SetStateAction } from "react";
+import { useCallback, useMemo, useState } from "react";
+
+import { DataSourceCategoryBrowser } from "@app/components/data_source_view/DataSourceCategoryBrowser";
+import { DataSourceNodeTable } from "@app/components/data_source_view/DataSourceNodeTable";
+import { CATEGORY_DETAILS } from "@app/lib/spaces";
+import {
+  useSpaceDataSourceViewsWithDetails,
+  useSpaces,
+} from "@app/lib/swr/spaces";
+import type {
+  ContentNodesViewType,
+  DataSourceViewType,
+  SpaceType,
+} from "@app/types";
+import type {
+  DataSourceViewCategoryWithoutApps,
+  DataSourceViewContentNode,
+  DataSourceViewSelectionConfigurations,
+  LightWorkspaceType,
+} from "@app/types";
+
+import { DataSourceSpaceSelector } from "./DataSourceSpaceSelector";
+
+type NavigationHistoryEntryType =
+  | { type: "root" }
+  | { type: "space"; space: SpaceType }
+  | { type: "category"; category: DataSourceViewCategoryWithoutApps }
+  | { type: "node"; node: DataSourceViewContentNode };
+
+/**
+ * Hook helper to manage the navigation history in the DataSourceBuilderSelector
+ *
+ * Shape is `[root, space, category, ...node]`
+ * so we case use index to update specific values
+ */
+const useNavigationHistory = () => {
```

Comment:
`nit`: what about moving it to a util file?

Reply:
Felt a very one-shot type of hook. Pretty much like the couple of helper function at the bottom.